### PR TITLE
Fix #8778: Make Entity Type and SubType extensible enums

### LINKED ISSUE #8778: Make Entity Type and SubType extensible enums
https://github.com/Azure/azure-sdk-for-net/pull/8677#discussion_r345383221

### PR DESCRIPTION
**Pull request by AlexanderSher**

### CODE REVIEW

File: sdk/core/Azure.Core/tests/TestFramework/ClientTestFixtureAttribute.cs
Comment (pakrym) [inline sdk/core/Azure.Core/tests/TestFramework/ClientTestFixtureAttribute.cs]: What's the point of this change?

Comment (AlexanderSher) [reply sdk/core/Azure.Core/tests/TestFramework/ClientTestFixtureAttribute.cs]: When `serviceVersions` is empty, `ProcessTest` marks all tests as "SkipRecordings".

Comment (pakrym) [reply sdk/core/Azure.Core/tests/TestFramework/ClientTestFixtureAttribute.cs]: Hm, but aren't both `serviceVersionNumber` and  `_maxServiceVersion` null in that case?

Comment (AlexanderSher) [reply sdk/core/Azure.Core/tests/TestFramework/ClientTestFixtureAttribute.cs]: Do you mean that checking `serviceVersion != null` should be enough? Let me check

Comment (pakrym) [reply sdk/core/Azure.Core/tests/TestFramework/ClientTestFixtureAttribute.cs]: I wonder why current condition doesn't work

Comment (AlexanderSher) [reply sdk/core/Azure.Core/tests/TestFramework/ClientTestFixtureAttribute.cs]: Current condition uses `int` instead of `int?`, so `serviceVersionNumber` is `0` and `_maxServiceVersion` is `null`.

Comment (AlexanderSher) [reply sdk/core/Azure.Core/tests/TestFramework/ClientTestFixtureAttribute.cs]: And `Convert.ToInt32` returns `0` for `null`.

Comment (AlexanderSher) [reply sdk/core/Azure.Core/tests/TestFramework/ClientTestFixtureAttribute.cs]: Ok, I've minimized the change to `ProcessTest` method itself.

File: sdk/textanalytics/Azure.AI.TextAnalytics/tests/TextAnalyticsRecordedTestSanitizer.cs
Comment (pakrym) [inline sdk/textanalytics/Azure.AI.TextAnalytics/tests/TextAnalyticsRecordedTestSanitizer.cs]: Do everyone use the same endpoint when recording?

Comment (AlexanderSher) [reply sdk/textanalytics/Azure.AI.TextAnalytics/tests/TextAnalyticsRecordedTestSanitizer.cs]: AFAIK, it is only used in `Azure.AI.TextAnalytics`

Comment (maririos) [reply sdk/textanalytics/Azure.AI.TextAnalytics/tests/TextAnalyticsRecordedTestSanitizer.cs]: Not really, I've seen Alexander and Anne using a different one than mine.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntitySubType.cs
Comment (maririos) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntitySubType.cs]: `Equals(object obj)` should be the one with `[EditorBrowsable(EditorBrowsableState.Never)]`

File: sdk/textanalytics/Azure.AI.TextAnalytics/tests/TextAnalyticsRecordedTestSanitizer.cs
Comment (pakrym) [reply sdk/textanalytics/Azure.AI.TextAnalytics/tests/TextAnalyticsRecordedTestSanitizer.cs]: This kind of sanitization might not work then. If in playback more the client would always use `https://test.cognitiveservices.azure.com/` but recordings would have something else they wouldn't match.

Comment (AlexanderSher) [reply sdk/textanalytics/Azure.AI.TextAnalytics/tests/TextAnalyticsRecordedTestSanitizer.cs]: I think it should work in combination with `SanitizeUri`:
```cs
public override string SanitizeUri(string uri)
{
    var index = uri.IndexOf(".cognitiveservices.azure.com", StringComparison.Ordinal);
    return index >= 0 ? "https://test" + uri.Substring(index) : base.SanitizeUri(uri);
}
```

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntitySubType.cs
Comment (AlexanderSher) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntitySubType.cs]: Fixed

File: sdk/textanalytics/Azure.AI.TextAnalytics/tests/TextAnalyticsRecordedTestSanitizer.cs
Comment (pakrym) [reply sdk/textanalytics/Azure.AI.TextAnalytics/tests/TextAnalyticsRecordedTestSanitizer.cs]: But why do we need this complication?

Comment (AlexanderSher) [reply sdk/textanalytics/Azure.AI.TextAnalytics/tests/TextAnalyticsRecordedTestSanitizer.cs]: I prefer not to expose any private data if I can. If @annelo-msft is ok with her cognitive services URL prefix on github, I can remove this sanitization.

Comment (pakrym) [reply sdk/textanalytics/Azure.AI.TextAnalytics/tests/TextAnalyticsRecordedTestSanitizer.cs]: My worry is that if the full uri is ever returned as part of the body (for example in paged responses) you won't be able to normalize it.

Comment (maririos) [reply sdk/textanalytics/Azure.AI.TextAnalytics/tests/TextAnalyticsRecordedTestSanitizer.cs]: >If @annelo-msft is ok with her cognitive services URL prefix on github

this will change according to who is running the records. If it is me, I use my own resource.
It might be someone external with their own resource too.
From what Pavel is saying, it looks like for now it might not be needed

**Review by maririos — APPROVED** [review]
Besides removing the need to sanitize `TEXT_ANALYTICS_ENDPOINT`, LGTM

File: sdk/textanalytics/Azure.AI.TextAnalytics/README.md
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/README.md]: Shouldn't need to call ToString() on this now?

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntitySubType.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntitySubType.cs]:
```suggestion
        /// Specifies that a DateTime named entity represents a date (e.g.: March 1, 1999).
```

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntitySubType.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntitySubType.cs]:
```suggestion
        /// Specifies that a DateTime named entity represents a time (e.g.: 12:34).
```

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntitySubType.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntitySubType.cs]:
```suggestion
        /// Specifies that a DateTime named entity represents a duration (e.g.: 2 minutes).
```

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntitySubType.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntitySubType.cs]:
```suggestion
        /// Specifies that a numeric type named entity is a number (e.g.: 5, six).
```

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntitySubType.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntitySubType.cs]:
```suggestion
        /// Specifies that a named entity of a Numeric type is a Percentage (e.g.: 50%, fifty percent).
```

Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntitySubType.cs]: Could we use this wording as the pattern across all the subtype docstring comments?

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntityType.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntityType.cs]:
```suggestion
        /// Specifies that the named entity corresponds to a Person.
```

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntityType.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntityType.cs]:
```suggestion
        /// Specifies that named entity contains natural or human-made landmarks, structures, or geographical features.
```

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntityType.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntityType.cs]:
```suggestion
        /// Specifies that the named entity contains the name of an organization, corporation, agency, or other group of people.
```

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntityType.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntityType.cs]:
```suggestion
        /// Specifies that the named entity contains a date, time or duration.
```

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntityType.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntityType.cs]:
```suggestion
        /// Specifies that the named entity contains a phone number (US Phone numbers only).
```

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntityType.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntityType.cs]:
```suggestion
        /// Specifies that the named entity contains an email address.
```

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntityType.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntityType.cs]:
```suggestion
        /// Specifies that the named entity contains an Internet URL.
```

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntityType.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntityType.cs]: Let's change this to URI on the client side, doing whatever needs to be done to correctly serialize it from the service value.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntityType.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntityType.cs]:
```suggestion
        /// Specifies that the named entity contains a number or numeric quantity.
```

File: sdk/textanalytics/Azure.AI.TextAnalytics/tests/TextAnalyticsRecordedTestSanitizer.cs
Comment (annelo-msft) [reply sdk/textanalytics/Azure.AI.TextAnalytics/tests/TextAnalyticsRecordedTestSanitizer.cs]: Correct me if I'm wrong, but doesn't the ARM work Heath was driving make it so we can always point to the same resource?  That way community members can run our tests without needed access to internal subscriptions?

File: sdk/textanalytics/Azure.AI.TextAnalytics/tests/samples/Sample4_RecognizeEntities.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/tests/samples/Sample4_RecognizeEntities.cs]: I think you don't need ToString here.

File: sdk/textanalytics/Azure.AI.TextAnalytics/README.md
Comment (AlexanderSher) [reply sdk/textanalytics/Azure.AI.TextAnalytics/README.md]: We still haven't decided about making `entity.SubType` a nullable struct here.

File: sdk/textanalytics/Azure.AI.TextAnalytics/tests/TextAnalyticsRecordedTestSanitizer.cs
Comment (AlexanderSher) [reply sdk/textanalytics/Azure.AI.TextAnalytics/tests/TextAnalyticsRecordedTestSanitizer.cs]: It is not related to the ARM template, it is about having the URL that was used in recorded session. But since we're not doing this type of sanitization anywhere else, I can remove it.

Comment (annelo-msft) [reply sdk/textanalytics/Azure.AI.TextAnalytics/tests/TextAnalyticsRecordedTestSanitizer.cs]: Will my endpoint go in recordings then?

Comment (pakrym) [reply sdk/textanalytics/Azure.AI.TextAnalytics/tests/TextAnalyticsRecordedTestSanitizer.cs]: Yes, that's how usually do it.

**Review by annelo-msft — APPROVED** [review]

**Comment by maririos** [comment]
@AlexanderSher do you have an estimate on when this PR can be merged?

**Comment by annelo-msft** [comment]
He's blocked on me right now @maririos ... looking ...

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntitySubType.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntitySubType.cs]: Why not just have this be `NamedEntitySubType("None")` ?

I'm concerned about users having to know they have to call ToString() vs. check for null.  The type is supposed to behave like an enum, and as I understand, the convention in .NET is to have a None value, and there's no concept of a null option in the collection of values.  But please lmk if I'm missing a reason to do it this way instead.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntitySubType.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/NamedEntitySubType.cs]: If we have just "None" (per comment above), we would just need `ToString() => _value;`

**Comment by annelo-msft** [comment]
Ok, if @AlexanderSher agrees with my comments, we should be good to merge this.

File: sdk/textanalytics/Azure.AI.TextAnalytics/README.md
Comment (annelo-msft) [reply sdk/textanalytics/Azure.AI.TextAnalytics/README.md]: It isn't that the change should be reverted, it's that you should be able to do `SubType: {entity.SubType}` like we do with entity.Type before it.

**Pull request merged**
